staticpolicy:smtalign: count for pre-allocated cpus for container

Tal-or opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

When CPUManager is configured with full-pcpus-only option,
it validates the number of available CPUs on the system.

There's a subtle bug in this check, where after kubelet restart
the containers already assigned with CPUs because
kubelet reads the checkpoint file and builds the container's
assignment map on startup.

We are not counting for these pre-allocated CPUs,
which could lead to false-positive, and prevent
the pod from being admitted.




#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes # https://github.com/kubernetes/kubernetes/issues/129078

#### Special notes for your reviewer:
N/A

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
N/A
```

Tal-or committed:
state:log: print CPUs in the assignment

`klog.InfoS` doesn't print the internal map of the container assigments.
The expected output is something like:

[<pod-uid>:[<container-name>:<cpus>]]

but <cpus> always printed as empty '{}'

Changing to `klog.Infof` solved the issue, and now
<cpus> contains meaningful value.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
staticpolicy:smtalign: count for pre-allocated cpus for container

When CPUManager is configured with full-pcpus-only option,
it validates the number of available CPUs on the system.

There's a subtle bug in this check, where after kubelet restart
the containers already assigned with CPUs because
kubelet reads the checkpoint file and builds the container's
assignment map on startup.

We are not counting for these pre-allocated CPUs,
which could lead to false-positive, and prevent
the pod from being admitted.

so let's move the code that checks if the container is already present in the state
to run BEFORE we check the SMT alignment and retrun early if
the contaier is already present.

This helps because in the SMT alignment flow, the
\`availablePhysicalCPUs\` value does not count for
pre-allocated CPUs (for containers which their CPUs were already
retrieved from the CPUManager checkpoint file).
so it might hit false-positive in the
\`if numCPUs > availablePhysicalCPUs\` check.

in addition:
Checking the checkpoint state before SMT alignment validation means that
existing allocations are preserved even if they violate the current policy.

For example, if kubelet previously ran with full-pcpus-only=false and is
restarted with full-pcpus-only=true, checkpointed allocations that don't
satisfy SMT alignment are accepted to avoid disrupting running workloads.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

ffromani commented on pkg/kubelet/cm/cpumanager/state/state_mem.go:
```diff
@@ -93,7 +93,7 @@ func (s *stateMemory) SetCPUAssignments(a ContainerCPUAssignments) {
 
 	s.assignments = a.Clone()
-	klog.InfoS("Updated CPUSet assignments", "assignments", a)
+	klog.Infof("Updated CPUSet assignments. assignments=%v", a)
```
while I seen the behavior myself, I'd rather fix `InfoS` somehow than drop it, especially because we want to move towards structured logging

Tal-or replied:
Reverted back to `InfoS` and added a `String()` function instead

ffromani commented:
/sig node
/ok-to-test
/triage accepted
/priority important-longterm

Tal-or commented:
I'm getting the following error from the linter:
```
ERROR: pkg/kubelet/cm/cpumanager/state/state_mem.go:96:2: unstructured logging function "Infof" should not be used (logcheck)
ERROR: 	klog.Infof("Updated CPUSet assignments. assignments=%v", a)
```
OTOH, switching back to `klog.InfoS` omits some important data:
see: https://github.com/kubernetes/kubernetes/pull/129079/commits/43c025329033f215c446f80b6714b7dfeeb266c4

Tal-or commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -354,6 +354,12 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
 
+		// availablePhysicalCPUs should consider guaranteed CPUs that already allocated to the Container.
+		// This might happen after a kubelet restart, for example.
+		if allocate, exists := s.GetCPUSet(string(pod.UID), container.Name); exists {
+			availablePhysicalCPUs += allocate.Size()
```
`GetAvailablePhysicalCPUs` called only in this context so it's possible to fix it with affecting other code areas.
But maybe `GetAvailablePhysicalCPUs`  assuming to start from a clean state is correct.
We should ask the original author :)

ffromani commented:
> I'm getting the following error from the linter:
> 
> ```
> ERROR: pkg/kubelet/cm/cpumanager/state/state_mem.go:96:2: unstructured logging function "Infof" should not be used (logcheck)
> ERROR: 	klog.Infof("Updated CPUSet assignments. assignments=%v", a)
> ```
> 
> OTOH, switching back to `klog.InfoS` omits some important data: see: [43c0253](https://github.com/kubernetes/kubernetes/commit/43c025329033f215c446f80b6714b7dfeeb266c4)

moving back to `Infof` is not an option indeed. Let's rather fix the `InfoS` call.

Tal-or commented on pkg/kubelet/cm/cpumanager/state/state.go:
```diff
@@ -29,7 +29,7 @@ func (as ContainerCPUAssignments) Clone() ContainerCPUAssignments {
 		ret[pod] = make(map[string]cpuset.CPUSet, len(as[pod]))
 		for container, cset := range as[pod] {
-			ret[pod][container] = cset
+			ret[pod][container] = cset.Clone()
```
After rethinking, there's no actual value here since `cpuset.CPUSET` is  immutable so we can't actually change the inner map.

ffromani replied:
Right, good point

Tal-or commented on pkg/kubelet/cm/cpumanager/policy_static_test.go:
```diff
@@ -636,15 +694,18 @@ func runStaticPolicyTestCase(t *testing.T, testCase staticPolicyTest) {
-	policy, _ := NewStaticPolicy(testCase.topo, testCase.numReservedCPUs, cpus, tm, testCase.options)
+	policy, err := NewStaticPolicy(testCase.topo, testCase.numReservedCPUs, cpus, tm, testCase.options)
+	if err != nil {
+		t.Errorf("failed to initialized StaticPolicy error (%v)", err)
```
Done. Thanks!

swatisehgal commented on pkg/kubelet/cm/cpumanager/policy_static_test.go:
```diff
@@ -564,6 +564,64 @@ func TestStaticPolicyAdd(t *testing.T) {
+			stDefaultCPUSet: cpuset.New(0, 5, 6, 11),
+			expErr:          nil,
+			expCPUAlloc:     true,
+			expCSet:         cpuset.New(2, 3, 8, 9),
```
We end up ultimately calling `Allocate` for the first container only and check against expected cpuset, so I don't really understand the benefit of having this test case: a pod with two containers.

I feel that we should either check if the allocation was as expected for the second container too or just have a single container in the test. 

Am I missing something?

Tal-or replied:
The purpose was to show that the flow behaves as expected for multiple containers as well as for a single one.

ffromani replied:
this is technically correct but makes the test hard to understand. We need either to drop the case per @swatisehgal suggestion, or add a new Test function which call `Allocate` for all containers, or add a comment that explains how the presence of other containers influence the call to the Allocate() against the first one only (e.g. how others are consumed by side effect, if at all, which is exactly the part hard to grasp).

Tal-or replied:
This was implement a long time ago. looking at it now again doesn't seem reasonable indeed. Removing

swatisehgal commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -354,6 +354,12 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
+		// availablePhysicalCPUs should consider guaranteed CPUs that already allocated to the Container.
+		// This might happen after a kubelet restart, for example.
+		if allocate, exists := s.GetCPUSet(string(pod.UID), container.Name); exists {
+			availablePhysicalCPUs += allocate.Size()
```
Adding size of allocated CPUs to `availablePhysicalCPUs` doesn't seem right to me. I can understand why is allows us to resolve the issue but it feels like we are applying a band-aid fix to solve the problem.

`availablePhysicalCPUs` is count of `defaultCPUSet` - `reservedPhysicalCPUs` as can be seen here:
https://github.com/kubernetes/kubernetes/blob/0d452a0e5b97cd66a308ccb5157ba989fa2a878e/pkg/kubelet/cm/cpumanager/policy_static.go#L281-L283

When CPUs are allocated, the `defaultSet` reduces subsequently reducing the number of physical CPUs available for allocation on that node (represented by `availablePhysicalCPUs`). If  we have observed from the state that the container already has CPUs it means that it was previously allocated CPUs and therefore the fix should be to skip allocation. Adding those pre-allocated CPUs back so they can be reallocated seems wrong. 


> this makes me realize that the problem could actually be that GetAvailablePhysicalCPUs assume to start from empty state; but after a kubelet restart the state is likely non-empty and this messes up the computation. 

I don't think that is the case here because `GetAvailablePhysicalCPUs` _correctly_ does not include CPU count corresponding to the containers that previously went running and whose state was stored in the checkpoint file.

Tal-or replied:
> If we have observed from the state that the container already has CPUs it means that it was previously allocated CPUs and therefore the fix should be to skip allocation. Adding those pre-allocated CPUs back so they can be reallocated seems wrong.

This situation here is quite unique, because the bug occurred when the `Allocate` flow called after kubelet restart.
It's true that the containers already had their cpus restored from the checkpoint file, but we are not sure that this fact is good enough for skipping the allocate flow.

Although the main allocate's flow purpose is allocating cpus to the containers, it might be doing additional work that if we'll choose to skip in certain situations (like kubelet restart), we might hit other problems we haven't anticipate.

swatisehgal replied:
Thanks @Tal-or but I still have concerns regarding the approach of adding pre-allocated CPUs back to `availablePhysicalCPUs`. My primary concern is that `availablePhysicalCPUs` should reflect count corresponding to only the truly available CPUs, which does not include CPUs already assigned to running containers.

> Although the main allocate's flow purpose is allocating cpus to the containers, it might be doing additional work that if we'll choose to skip in certain situations (like kubelet restart), we might hit other problems we haven't anticipate.

This is a bit speculative, and it would be good to at least explore skipping allocation as an alternative path. Let's clearly understand if there are unintended implication of taking this path. 

Skipping a pod when it is already in the state is not something new, I have already seen being done in CPU Manager. Here is an example:

https://github.com/kubernetes/kubernetes/blob/ea73dad54981c923c03494348104f125639dbdb0/pkg/kubelet/cm/cpumanager/policy_static.go#L370-L374 

Can we look into the code to have a definitive view on this? If that is not a suitable path its okay but it would be good to understand why that is the case.

Tal-or replied:
The only concern I've with this approach is as follow:
t1 - container `A` assigned to an odd number of CPUs
t2 - user configured SMT alignment for kubelet
t3 - kubelet got restart to enable SMT alignment 
t4 - container `A` will slip through because it is present in the state and now we return before the SMT alignment check.

swatisehgal reviewed (commented):
Thanks for your work on this. It is a great catch but I feel that the implementation needs some more thought. I have suggested an alternative approach.

swatisehgal reviewed (commented):
I'd like us to also account for some additional edge cases that could occur in production environments. In particular, we should validate the behavior when both the SMT policy options AND the reserved CPU configuration are changed between kubelet restarts.

My key concern is: what happens if a CPU previously allocated to a container becomes part of the reserved CPU set after a restart?
For example:

```
# Before restart:
--reserved-cpus=0,1
--cpu-manager-policy-options=full-pcpus-only=true
# Container has CPUs: 2,3,8,9

# After restart:  
--reserved-cpus=0,1,2,3  # CPUs 2,3 now reserved!
--cpu-manager-policy-options=full-pcpus-only=true
# Existing allocation conflicts with new reservation
```

This scenario could lead to resource conflicts if the allocation of previously allocated containers is preserved. We likely need an additional validation step that checks if existing allocations conflict with the current configuration, and if so, logs a warning while deciding whether to preserve or reallocate. How we handle this case might need further community discussion. 

To ensure we're covered, let’s add some targeted e2e tests for the following cases:

1. Primary Fix Validation:
Deploy pod with full-pcpus-only=false
Restart kubelet (config changes to enable SMT policy option)
Verify container maintains original CPU allocation

1. Reserved CPU Conflict Scenario:
Deploy pod with full-pcpus-only=true
Deploy pod with allocated CPUs (e.g., 2,3,8,9)
Restart kubelet with expanded reserved CPU set including previously allocated CPUs and Deploy pod with full-pcpus-only=true
Verify behavior: Should the allocation be preserved or reallocated?

1. Mixed Configuration Changes:
Test combinations of SMT policy option + reserved CPU changes
Ensure system remains stable and logs appropriate warnings

swatisehgal commented:
> Hey @swatisehgal, I appreciate the thorough coverage that you were suggested but there's already a validation for this use case: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/policy_static.go#L222

Thanks for pointing me to the validation code for this. Do we have e2e tests for the scenarios I have captured above as well? If not, can we add them to validate your fix and ensure that we are not causing any regressions? 

> In addition this code flow/checks were not meant to deal with Kubelet configuration changes. A validation of configuration changes should be made at an early phase of Kubelet stratup rather than in the Allocate() flow of the container.
I understand but the issue we are trying to fix becomes evident because kubelet configuration is updated to full-pcpus-only-option to true followed by a restart. What I am suggesting above is that along with the policy option config change we need to make sure that the scenario where update to the reserved CPUs (such that they are overlapping with already allocated CPUs) is handled appropriately.

Tal-or commented:
>I understand but the issue we are trying to fix becomes evident because kubelet configuration is updated to full-pcpus-only-option to true followed by a restart. 

That's my point exactly, the issue is not related to configuration update.
We're talking about a plain kubelet restart here, with no configuration changes.

com30n commented:
> I'm still not sure you are hitting this behavior or a different one.

It looks like this, but I'm still not sure I understand why kubelet wants to create a new pod with a new cpuset for the static pod apiserver if the static pod already exists.

I left a comment in the original issue with my reproduction and steps on how to get the same behavior.

com30n commented:
> on the other hand the feature is expected to be used in worker nodes which don't run control plane components

btw, I also discovered that pods with tasksets on worker nodes also encounter this issue if the node was fully loaded and/or the node did not have enough cpus in the default cpuset.

Thus, kubelet will simply terminate all pods with tasksets that request more cpu than are freely available in the default cpuset.

So, restarting kubelet may result in restarting all pods under the cpumanager.

Tal-or committed:
e2e:node: consider pre-allocated CPUs

This test verifies that pods with pre-allocated CPUs (from the checkpoint file)
are not rejected after kubelet restart when SMT alignment is enabled.
Regression test for the fix where the container presence check was moved
before the SMT alignment check.

The key is to request enough CPUs so that if pre-allocated CPUs are not
counted, the SMT alignment check would fail due to insufficient available
physical CPUs.

Calculate the maximum SMT-aligned CPUs we can request
We need to request most of the allocatable CPUs to trigger the bug.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

ffromani reviewed (commented):
thanks for the updates. It seems we're converging but I'd like a bit more details in some areas before we can merge. Looking good though.

ffromani reviewed (commented):
Thinking more about it, https://github.com/kubernetes/kubernetes/pull/129079/files#r2698744928 is the better approach over the targeted fix because the current code silently broke a KEP promise, and now we're aligning again with the KEP intent.

If you resubmit, please inline (preferred) or at least make private `GetAvailablePhysicalCPUs`. It is used once and has no known consumer outside this package, so we should default to private.

Tal-or committed:
staticpolicy: inline `GetAvailablePhysicalCPUs`

The function used once and has no known consumer outside the package.
see: https://github.com/kubernetes/kubernetes/pull/129079\#pullrequestreview-3691692701

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or commented:
/retest-required

ffromani commented on pkg/kubelet/cm/cpumanager/state/state.go:
```diff
@@ -37,6 +39,46 @@ func (as ContainerCPUAssignments) Clone() ContainerCPUAssignments {
+	for i, pod := range pods {
+		containerMap := as[pod]
+		fmt.Fprintf(&sb, "%q:{", pod)
+
+		// Sort containers alphabetically
+		containers := make([]string, 0, len(containerMap))
+		for container := range containerMap {
+			containers = append(containers, container)
+		}
+		sort.Strings(containers)
+
+		for j, container := range containers {
+			fmt.Fprintf(&sb, "%q:%q", container, containerMap[container].String())
+			if j < len(containers)-1 {
+				sb.WriteString(",")
+			}
+		}
+		sb.WriteString("}")
+		if i < len(pods)-1 {
+			sb.WriteString(",")
+		}
+	}
+	sb.WriteString("}")
+	return sb.String()
+}
```
*IF* you happen to resubmit for other reasons, we can simplify:
```
func (as ContainerCPUAssignments) String() string {
        var sb strings.Builder
        sb.WriteString("{")

        for i, pod := range slices.Sorted(maps.Keys(as)) {
                containerMap := as[pod]
                fmt.Fprintf(&sb, "%q:{", pod)

                for j, container := range slices.Sorted(maps.Keys(containerMap)) {
                        fmt.Fprintf(&sb, "%q:%q", container, containerMap[container].String())
                        if j < len(containers)-1 {
                                sb.WriteString(",")
                        }
                }
                sb.WriteString("}")
                if i < len(pods)-1 {
                        sb.WriteString(",")
                }
        }
        sb.WriteString("}")
        return sb.String()
}
```

Tal-or replied:
Done.

ffromani commented on test/e2e_node/cpu_manager_test.go:
```diff
@@ -1009,6 +1009,69 @@ var _ = SIGDescribe("CPU Manager", ginkgo.Ordered, ginkgo.ContinueOnFailure, fra
+			ginkgo.By("ensuring kubelet is healthy after restart")
+			gomega.Eventually(ctx, func() bool {
+				return kubeletHealthCheck(kubeletHealthCheckURL)
+			}, f.Timeouts.PodStart, f.Timeouts.Poll).Should(gomega.BeTrueBecause("kubelet should be healthy after restart"))
```
if you resubmit: use `WithTimeout` and `WithPolling` in the new code.

Tal-or replied:
Done. Thanks

ffromani reviewed (commented):
/lgtm

prior to this PR, the admission flow was accounting for the physical CPUs twice on kubelet restart.
Twice because the code was 1. accounting for the physical CPU availability as normal, but 2. on top of the PRIOR accounting done when the pod was first admitted, and whose consumption was consolidated on the state file.
The double accounting prevented correct pods to run, but was confined to the admission flow only because just after the physical cpu availability check the code was checking if the pod was already present in the state file, and if so, short circuting out.
The bug therefore manifested itself only at restart and only with pods whose CPU consumption exceeded the node capacity free left minus its own capacity, which is a narrow set of cases.

there are quite some improvements we can make still, but the core and the extra improvements this PR delivers are sound, and we've waited quite a lot to get in.
Let's make progress.

ffromani commented:
/test pull-kubernetes-unit-windows-master

ffromani commented:
/test pull-kubernetes-unit-windows-master

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd

Tal-or commented:
@ffromani CI showed the following error:
`test/e2e_node/cpu_manager_test.go:1061:12: undefined: kubeletHealthCheck` 

`kubeletHealthCheck` function was removed so the PR required a rebase and reuse of a different function.

Tal-or commented:
/retest-required

ffromani commented:
/test pull-kubernetes-kind-dra-n-3
/test pull-kubernetes-kind-dra-n-2
/test pull-kubernetes-kind-dra-n-1
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-device-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-podlevelresources
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani reviewed (commented):
/approve
/lgtm

thanks!
